tests/test.rs

```diff
@@ -0,0 +1,105 @@
+use insta;
+use os_pipe::pipe;
+use std::io::Read;
+use std::sync::Mutex;
+use structopt::lazy_static::lazy_static;
+
+lazy_static! {
+    /// rust will try to run the tests in parallel by default, and `insta` doesn't like the
+    /// usual way of disabling this via `--test-threads=1`, so we're using this
+    /// mutex to make sure we're not re-flashing until the last run is finished
+    static ref ONE_RUN_AT_A_TIME: Mutex<i32> = Mutex::new(0i32);
+}
+
+/// run probe-run with `args` and truncate the "Finished .. in .." and "Running `...`" flashing output
+/// NOTE: this currently only capures `stdin`, so any `log::` ed output, like flashing
+fn run_and_truncate(args: &str) -> String {
+    let _guard = ONE_RUN_AT_A_TIME.lock().unwrap();
+
+    // add prefix to run this repository's version of `probe-run` and
+    // remove user-dependent registry and rustc information from backtrace paths
+    let complete_command = format!("run -- {} --shorten-paths", args);
+    let complete_command_args = complete_command.split(" ");
+
+    let (mut reader, writer) = pipe().unwrap();
+    let writer_clone = writer.try_clone().unwrap();
+
+    let mut command = std::process::Command::new("cargo");
+    command.args(complete_command_args);
+
+    // capture stderr and stdout while preserving line order
+    command.stdout(writer);
+
+    command.stderr(writer_clone);
+
+    // run `probe-run`
+    let mut handle = command.spawn().unwrap();
+
+    // force EOF
+    drop(command);
+
+    // retrieve output and clean up
+    let mut probe_run_output = String::new();
+    reader.read_to_string(&mut probe_run_output).unwrap();
+    handle.wait().unwrap();
+
+    // remove the lines printed during flashing, as they contain timing info that's not always the same
+    let mut truncated_probe_run_output = "".to_string();
+    for line in probe_run_output.lines() {
+        if !line.starts_with("    Finished")
+            && !line.starts_with("     Running `")
+            && !line.starts_with("    Blocking waiting for file lock ")
+            && !line.starts_with("   Compiling probe-run v")
+            // TODO don't drop these after https://github.com/knurling-rs/probe-run/issues/217 is resolved
+            && !line.starts_with("└─ ")
+        {
+            truncated_probe_run_output.push_str(line);
+            truncated_probe_run_output.push_str("\n");
+        }
+    }
+
+    truncated_probe_run_output
+}
+
+#[test]
```

Comment:
Could `rstest` be used to combine these tests into one with various cases? See https://docs.rs/rstest/0.10.0/rstest/#creating-parametrized-tests

Reply:
(this obviously can be done in another PR)

Reply:
mhm, style-wise I personally don't think that would be a good idea for the *current* test cases, as they all check different functionalities and are pretty short. We'd be losing descriptiveness (test names as documentation) and wouldn't really win much in terms of deduplicating code. In other cases it might be  super useful though, e.g. if we're testing different values for `--max-backtrace-len`?

I'm not sure if/how that would work well with `insta`, as the snapshots carry the names of the test functions, but maybe that can be tweaked or insta is nice enough to encode any additional params in the name? Would be great to investigate in that future PR :)

Reply:
👍🏾